### ISSUE TITLE
xxx has incomplete type and cannot be defined (cython doesn't warn the user)

### ISSUE THREAD

**Issue by gabrieldemarmiesse** [cython/cython#2098]
Hi! I'm new to Cython, and I've encountered a bug at the compiler level. I must do something wrong, but it's difficult to track since it's in the generated c++ file. Maybe Cython can throw an error in this case?

I tried to make a minimal example so that it's easier to understand ( and possible help me).
I'm trying to use a c++ class in Cython. 

I use Windows and Anaconda with python 3.6.
With `pip freeze`, I get `Cython==0.27.3`

#### the `wrapper.pxd` file:
```python
cdef extern from "arithmetic_codec.h" nogil:

    cdef cppclass Static_Data_Model:
        Static_Data_Model()
        unsigned model_symbols()
        void set_distribution(unsigned number_of_symbols,
                              const double probability[] = 0)

```

#### The `no_context.pyx` file:
```python
# distutils: sources = ecoding/arithmetic_codec.cpp

import numpy as np
from wrapper cimport *


cpdef _compile(self):
    cdef Static_Data_Model data_model

    distribution = np.array([4,3,8])
    data_model = Static_Data_Model()

    cdef double [::1] distribution_view = distribution
    cdef unsigned number_of_symbols = len(distribution)
    cdef double* my_ptr = &distribution_view[0]
    data_model.set_distribution(number_of_symbols, my_ptr)
```

#### the `setup.py` file:
```python
from setuptools import setup
from setuptools import find_packages
from distutils import core
from Cython.Build import cythonize

setup(name='Entropy-Coding',
      version='0.2',
      install_requires=['numpy', 'Cython'],
      packages=find_packages())

core.setup(
    ext_modules=cythonize("ecoding/no_context.pyx",
                          annotate=True,
                          language='c++')
)
```

#### Here is the error message:
```
python setup.py build_ext --inplace
running build_ext
Please put "# distutils: language=c++" in your .pyx or .pxd file(s)
running build_ext
building 'ecoding.no_context' extension
C:\x86_64-5.4.0-release-posix-seh-rt_v5-rev0\mingw64\bin\gcc.exe -mdll -O -Wall -DMS_WIN64 -I.\ecoding -IC:\Users\smith\Miniconda3\include -IC:\Users\smith\Miniconda3\include -c ecoding/no_context.cpp -o build\temp.win-amd64-3.6\Release\ecoding\no_context.o
ecoding/no_context.cpp: In function 'PyObject* __pyx_f_7ecoding_10no_context__compile(PyObject*, int)':
ecoding/no_context.cpp:1894:62: error: aggregate '__pyx_f_7ecoding_10no_context__compile(PyObject*, int)::__pyx_opt_args_17Static_Data_Model_set_distribution __pyx_t_10' has incomplete type and cannot be defined
   struct __pyx_opt_args_17Static_Data_Model_set_distribution __pyx_t_10;
                                                              ^
error: command 'C:\\x86_64-5.4.0-release-posix-seh-rt_v5-rev0\\mingw64\\bin\\gcc.exe' failed with exit status 1
```

Thank you for the help.

**Comment by gabrieldemarmiesse** [comment]
After further research, it seems that it's the default argument causing the error. It's strange though that Cython doesn't warn about it.

**Comment by Alexhuszagh** [comment]
I'm not a Cython core contributor or maintainer, but I'm going to chime in my two cents. What would you like Cython to do? To my knowledge, Cython doesn't do any C/C++ parsing (the latter is a complex task), but rather uses the `extern cdef` to generate C++ code with the appropriate includes that can interface with the existing class or function.

That's why the following example works:

**factorial.h**

```cpp
unsigned factorial(unsigned x = 1)
{
    if (x == 0) { 
        return 1;
    }
    return(x * factorial(x - 1));
}
```

**factorial.pxd**

```cython
cdef extern from "factorial.h" nogil:
    cdef unsigned factorial()
    cdef unsigned factorial(unsigned x)
```

In short, Cython doesn't know anything about whether the type is complete or not: the C/C++ compiler it uses to compile the generated source files, however, does. And in this case it gives you the exact warning (the type is not complete).

**Comment by gabrieldemarmiesse** [comment]
Thank you very much for this clear explanation. My mistake was to assume that Cython would parse the C++ header file.
